indexer/string_slice.cpp

```diff
@@ -0,0 +1,74 @@
+#include "indexer/string_slice.hpp"
+
+namespace search
+{
+JoinIterator::JoinIterator(StringSliceBase const & slice, Position position) : m_slice(slice)
+{
+  if (position == Position::Begin)
+  {
+    m_string = 0;
+    m_offset = 0;
+    Normalize();
+  }
+  else
+  {
+    m_string = GetMaxSize();
+    m_offset = 0;
+  }
+}
+
+// static
+JoinIterator JoinIterator::Begin(StringSliceBase const & slice)
+{
+  return JoinIterator(slice, Position::Begin);
+}
+
+// static
+JoinIterator JoinIterator::End(StringSliceBase const & slice)
+{
+  return JoinIterator(slice, Position::End);
+}
+
+JoinIterator & JoinIterator::operator++()
+{
+  ++m_offset;
+  Normalize();
+  return *this;
+}
+
+void JoinIterator::Normalize()
```

Comment:
No, this is the normalization of iterator, i.e. skipping of empty strings. I can rename it to SkipEmptyStrings.